RPC refactoring: Access wallet using new GetWalletForJSONRPCRequest

luke-jr opened the pull request:
Part of the refactorings needed for basic multiwallet (#8694)

luke-jr committed:
RPC/Wallet: Pass CWallet as pointer to helper functions

luke-jr committed:
Move nWalletUnlockTime to CWallet::nRelockTime, and name timed task unique per CWallet

jonasschnelli commented:
I don't like the coupling and the `#ifdef ENABLE_WALLET` in `rpc/server.cpp|.h`.
I'd recommend to keep the `CRPCRequestInfo` wallet-free.

I think we should pack the request path (URI) into the `CRPCRequestInfo` and or informations about the authentication (in case we want to distinct wallets based on authentication).
Then I think there should be a method in `wallet.cpp` (or in `rpcwallet.cpp`) that maps a `CWallet *` pointer from a given URI, Auth-Info or the complete `CRPCRequestInfo` instance.

Instead of the `CWallet *& pwallet = reqinfo.wallet;` there could be then something like `CWallet *pwallet = CWallets::getWalletFromRequest(reqinfo)`

jonasschnelli commented:
General ConceptACK on a `CRPCRequestInfo` for the RPC table commands.
Maybe it could also include the `UniValue params` and `fHelp`?

paveljanik commented on src/rpc/misc.cpp:
```diff
@@ -70,7 +70,8 @@ UniValue getinfo(const UniValue& params, bool fHelp)
 
 #ifdef ENABLE_WALLET
-    LOCK2(cs_main, pwalletMain ? &pwalletMain->cs_wallet : NULL);
+    CWallet *& pwallet = reqinfo.wallet;
```
`*&`?

luke-jr replied:
Yes, pwallet is an alias to reqinfo.wallet which is of type CWallet*.

luke-jr commented:
> I think we should pack the request path (URI) into the CRPCRequestInfo and or informations about the authentication (in case we want to distinct wallets based on authentication).
Then I think there should be a method in wallet.cpp (or in rpcwallet.cpp) that maps a CWallet * pointer from a given URI, Auth-Info or the complete CRPCRequestInfo instance.

That sounds nice, but greatly complicates the implementation. :(

MarcoFalke commented:
I think this can be closed after #8788?

jonasschnelli commented:
Closing in favor of merged #8788

luke-jr committed:
RPC: Do all wallet access through new GetWalletForJSONRPCRequest

luke-jr committed:
Reformat touched lines with C++11

luke-jr commented:
Rebased and refactored based on @jonasschnelli 's idea.

laanwj commented:
Makes sense, utACK ab5ce98

luke-jr commented:
Rebased and addressed nit

luke-jr committed:
RPC: Pass on JSONRPCRequest metadata (URI/user/etc) for "help" method

instagibbs commented:
utACK 7de55733c5690350c48ec94660b5be56fbb5eb39

luke-jr commented:
Minor change: Forward-declared CWallet even for non-wallet builds so it can be used in a pointer type, avoiding unnecessary casting.

TheBlueMatt commented on src/rpc/rawtransaction.cpp:
```diff
@@ -35,6 +35,11 @@
 using namespace std;
 
+#ifdef ENABLE_WALLET
+std::string HelpRequiringPassphrase(CWallet *);
```
I'd kinda prefer we add another declaration outside of a header in an unrelated file...can we keep it in server.h and ifdef ENABLE_WALLET it?

luke-jr replied:
I suppose we could. But then it's more likely to get used in new code (which I think we want to discourage?)

TheBlueMatt replied:
I'd think we can (and, more importantly, would) still nag people submitting PRs which add more uses of it outside of src/wallet/rpc*.

TheBlueMatt commented on src/rpc/misc.cpp:
```diff
@@ -26,6 +26,10 @@
 using namespace std;
 
+#ifdef ENABLE_WALLET
+CWallet *GetWalletForJSONRPCRequest(const JSONRPCRequest&);
```
Here as well.

TheBlueMatt commented on src/wallet/rpcdump.cpp:
```diff
@@ -29,6 +29,7 @@
 
+CWallet *GetWalletForJSONRPCRequest(const JSONRPCRequest&);
 void EnsureWalletIsUnlocked(CWallet *);
 bool EnsureWalletIsAvailable(CWallet *, bool avoidException);
```
Can we move these to some rpc.h in src/wallet? Why are we declaring functions in another file from the definition if they're both in src/wallet???

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -30,6 +30,11 @@ using namespace std;
 int64_t nWalletUnlockTime;
 static CCriticalSection cs_nWalletUnlockTime;
 
+CWallet *GetWalletForJSONRPCRequest(const JSONRPCRequest& request)
```
Can you document this function somewhere (also probably move its definition to src/wallet/rpcwallet.h) - a few places in src/wallet assume it always returns non-NULL but this is not documented (I havent looked at the actual multi-wallet PR, but does this function then throw an RPC exception if you ask for a wallet that isnt loaded, or does it return NULL then)?

luke-jr committed:
Move wallet RPC declarations to rpcwallet.h

gmaxwell commented:
Concepts ACK.

So in some cases in the RPC you get the wallet pointer from json but then the check if it's available is far below.  This is begging for someone to insert code that uses a potentially null pointer between to two and doesn't notice because their function doesn't get tested with a missing wallet.  I would recommend moving the creation of that local pointer down to the point where you're going to test it.

Alternatively or in addition,  rename GetWalletForJSONRPCRequest to TryGetWalletForJSONRPCRequest and make GetWalletForJSONRPCRequest a wrapper for it that throws when it fails?

luke-jr committed:
More tightly couple EnsureWalletIsAvailable with GetWalletForJSONRPCRequest where appropriate

luke-jr commented:
I liked the TryGetWalletForJSONRPCRequest/GetWalletForJSONRPCRequest refactor idea, but I don't see any clean way to do it without changing the `help` behaviours.

So I moved the one case EnsureWalletIsAvailable was delayed up, and removed the blank line between them and GetWalletForJSONRPCRequest.

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2019,9 +2016,8 @@ UniValue walletpassphrase(const JSONRPCRequest& request)
-    nWalletUnlockTime = GetTime() + nSleepTime;
-    RPCRunLater("lockwallet", boost::bind(LockWallet, pwallet), nSleepTime);
+    pwallet->nRelockTime = GetTime() + nSleepTime;
+    RPCRunLater(strprintf("lockwallet_%u", uintptr_t(pwallet)), boost::bind(LockWallet, pwallet), nSleepTime);
```
I super dont like the fact that the pointer to the wallet will end up in debug.log if you have debug=rpc enabled.

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -340,8 +349,7 @@ UniValue getaddressesbyaccount(const JSONRPCRequest& request)
     UniValue ret(UniValue::VARR);
-    BOOST_FOREACH(const PAIRTYPE(CBitcoinAddress, CAddressBookData)& item, pwallet->mapAddressBook)
-    {
+    for (const PAIRTYPE(CBitcoinAddress, CAddressBookData)& item : pwallet->mapAddressBook) {
```
nit: I believe you can remove the PAIRTYPE and replace with std::pair since we're not using BOOST_FOREACH now.

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1219,8 +1241,7 @@ UniValue ListReceived(CWallet * const pwallet, const UniValue& params, bool fByA
     map<string, tallyitem> mapAccountTally;
-    BOOST_FOREACH(const PAIRTYPE(CBitcoinAddress, CAddressBookData)& item, pwallet->mapAddressBook)
-    {
+    for (const PAIRTYPE(CBitcoinAddress, CAddressBookData)& item : pwallet->mapAddressBook) {
```
nit: Here as well (no PAIRTYPE).

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1635,14 +1665,14 @@ UniValue listaccounts(const JSONRPCRequest& request)
     map<string, CAmount> mapAccountBalances;
-    BOOST_FOREACH(const PAIRTYPE(CTxDestination, CAddressBookData)& entry, pwallet->mapAddressBook) {
-        if (IsMine(*pwallet, entry.first) & includeWatchonly) // This address belongs to me
+    for (const PAIRTYPE(CTxDestination, CAddressBookData)& entry : pwallet->mapAddressBook) {
```
nit: Here as well (no PAIRTYPE).

TheBlueMatt reviewed (commented):
Please do not tag the 4th commit "Trivial". We usually define trivial as doesnt touch any code.

As for the 6th commit: please do not do this. nothing in src/wallet is built when ENABLE_WALLET is off, so generally src/wallet/* should never be included outside of src/wallet/*

Aside from the (partial-revert) of "Move wallet RPC declarations to rpcwallet.h" and the printing of the raw pointers to debug log, this looks good to me at d9aff6e322fd4cdba8d338d304a627af46feb7eb.

gmaxwell commented:
@luke-jr Looks good to me, will test as soon as you update for matt's latest comments.

luke-jr committed:
Wallet/RPC: Use filename rather than CWallet pointer, for lockwallet RPCRunLater job name

The job name is logged, and could pose as an information leak to someone attacking the process, helping them counteract ASLR protections

luke-jr committed:
Wallet: Sanitise -wallet parameter

luke-jr commented:
Made changes requested by @TheBlueMatt, and rebased to resolve conflict. Also includes a quick sanity check that -wallet doesn't include path separators.

TheBlueMatt commented:
While you're at it, can you call SanitizeString (in addition to the "/\" check) on the wallet param?

TheBlueMatt commented:
utACK 7d1228bf7f1c398d9fcda6161e6aebfc5e5e3700

gmaxwell commented:
Needs rebase.

gmaxwell commented:
ACK.

fanquake commented:
Needs rebasing again.

luke-jr commented:
Rebased yet again.

gmaxwell commented:
still ACK

BlockMechanic commented on src/wallet/wallet.cpp:
```diff
@@ -3740,6 +3740,12 @@ bool CWallet::InitLoadWallet()
 
     std::string walletFile = GetArg("-wallet", DEFAULT_WALLET_DAT);
```
Should this not be "count" i.e for multiple wallet files to be created ? Adding perhaps a for loop to iterate over required num of wallets?

luke-jr replied:
No, because this PR is not multiwallet...

ryanofsky commented on src/rpc/misc.cpp:
```diff
@@ -234,6 +235,9 @@ UniValue validateaddress(const JSONRPCRequest& request)
 }
 
+// Needed even with !ENABLE_WALLET, to pass (ignored) pointers around
+class CWallet;
```
This doesn't actually seem to be needed since CWallet is also forward declared in init.h, but maybe it is better to keep it.

ryanofsky reviewed (commented):
utACK d678771c660c936a6222dd43a204cb2699100f3e

jonasschnelli commented:
Great PR. I think we should get this in as soon as possible.
utACK d678771c660c936a6222dd43a204cb2699100f3e

laanwj commented:
utACK d678771

jtimon commented:
By the way, I reviewed this partially after merged when rebasing #9845 (making it mostly pointless since most was done here already).

adneerav commented:
Hello @luke-jr ,

Is there any way where we can increase the number of wallets to be in loaded state at same time.As rightnow its like if I load wallet more than 100 or 150 daemon is getting crashed/stopped. I am trying to load/unload wallet dynamically using RPC.

What i tried is incraesing the `set_lk_max_lockers` to 400000 and also added
`set_lg_regionmax 1048576` But still same things I am facing.

Reflink for above setup : https://docs.oracle.com/cd/E17276_01/html/api_reference/C/set_lk_max_lockers_parameter.html
https://bitcoin.stackexchange.com/a/70737

So is there any way or configuration or any changes which can increase and allow more number of wallets to be loaded ?
